### PR TITLE
Avoid intermediary array allocation in incremental generator

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
This drops us from 3GB of alloc to 2.3 GB (a good 25% reduction) during incremental generation.

Specifically it removes this array:

Before:
image

After:
image

### CODE REVIEW

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 1]
```diff
@@ -128,19 +128,15 @@ public NodeStateTable<T> AsCached()
             return (_states[^1].GetItem(0), HasTrackedSteps ? Steps[^1] : null);
         }
 
-        public ImmutableArray<NodeStateEntry<T>> Batch()
+        public IEnumerable<NodeStateEntry<T>> Batch()
         {
-            var sourceBuilder = ArrayBuilder<NodeStateEntry<T>>.GetInstance();
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:133]: this the array we're avoiding creating.

**Review by chsienki — APPROVED** [review]
Nice! LGTM

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 2]
```diff
@@ -128,19 +128,15 @@ public NodeStateTable<T> AsCached()
             return (_states[^1].GetItem(0), HasTrackedSteps ? Steps[^1] : null);
         }
 
-        public ImmutableArray<NodeStateEntry<T>> Batch()
+        public IEnumerable<NodeStateEntry<T>> Batch()
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:131]: sure.  let me try that.

**Comment by CyrusNajmabadi** [comment]
@cston much nicer. thanks!

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs [hunk 3]
```diff
@@ -49,9 +50,19 @@ public NodeStateTable<ImmutableArray<TInput>> UpdateStateTable(DriverStateTable.
...
 
-            var batchedSourceEntries = sourceTable.Batch();
-            var sourceValues = batchedSourceEntries.SelectAsArray(sourceEntry => sourceEntry.State != EntryState.Removed, sourceEntry => sourceEntry.Item);
-            var sourceInputs = newTable.TrackIncrementalSteps ? batchedSourceEntries.SelectAsArray(sourceEntry => (sourceEntry.Step!, sourceEntry.OutputIndex)) : default;
+            var sourceValuesBuilder = ArrayBuilder<TInput>.GetInstance();
+            var sourceInputsBuilder = newTable.TrackIncrementalSteps ? ArrayBuilder<(IncrementalGeneratorRunStep InputStep, int OutputIndex)>.GetInstance() : null;
+
+            foreach (var entry in sourceTable)
+            {
+                if (entry.State != EntryState.Removed)
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs:58]: >EntryState.Removed [](http://example.com/codeflow?start=35&length=18)

Do we need to check `sourceTable.HasTrackedSteps` as well?

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs [hunk 4]
```diff
@@ -49,9 +50,19 @@ public NodeStateTable<ImmutableArray<TInput>> UpdateStateTable(DriverStateTable.
...
+                    sourceValuesBuilder.Add(entry.Item);
+
+                sourceInputsBuilder?.Add((entry.Step!, entry.OutputIndex));
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs:61]: >sourceInputsBuilder?.Add [](http://example.com/codeflow?start=16&length=24)

Should adding to `sourceInputsBuilder` rely on the same conditions as `sourceValuesBuilder`?

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs:61]: no.  we always want to add the information about the steps *including* for a 'remove'.  i can add comments.

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs [hunk 5]
```diff
@@ -128,25 +127,8 @@ public NodeStateTable<T> AsCached()
...
-        {
-            var sourceBuilder = ArrayBuilder<NodeStateEntry<T>>.GetInstance();
-            foreach (var entry in this)
-            {
-                // If we have tracked steps, then we need to report removed entries to ensure all steps are in the graph.
-                // Otherwise, we can just return non-removed entries to build the next value.
-                if (entry.State != EntryState.Removed || HasTrackedSteps)
-                {
-                    sourceBuilder.Add(entry);
-                }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs:141]: @cston  the `HasTrackedSteps` was to make sure that we retruend this entry from the iterator so we could associated it with the perf data we're collecting.  On the receiver side though we then filtered this out from the actual entry-list we were creating, but we used it for the step-list being built.

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs [hunk 6]
```diff
@@ -49,9 +50,22 @@ public NodeStateTable<ImmutableArray<TInput>> UpdateStateTable(DriverStateTable.
...
+            var sourceValuesBuilder = ArrayBuilder<TInput>.GetInstance();
+            var sourceInputsBuilder = newTable.TrackIncrementalSteps ? ArrayBuilder<(IncrementalGeneratorRunStep InputStep, int OutputIndex)>.GetInstance() : null;
+
+            foreach (var entry in sourceTable)
+            {
+                // At this point, we can remove any 'Removed' items and ensure they're not in our list of states.
+                if (entry.State != EntryState.Removed)
+                    sourceValuesBuilder.Add(entry.Item);
+
+                // However, regardless of if the entry was removed or not, we still keep track of it's step information
```
Comment (cston) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs:62]: >it's [](http://example.com/codeflow?start=98&length=4)

Typo: "its"

**Review by cston — APPROVED** [review]

File: src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs:62]:
```suggestion
                // However, regardless of if the entry was removed or not, we still keep track of its step information
```

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/SourceGeneration/Nodes/BatchNode.cs]: fixed. t hanks!

**Pull request merged**